src/player/grab_rotation.c

```diff
@@ -0,0 +1,103 @@
+    return grabRotationFlagsForDecorId(decorIdForCollisionObject(collisionObject)); // this will probably need to be replaced in the future
+}
+
+
+void grabRotationApplySnapToCubeNormals(const enum GrabRotationFlags flags, struct Quaternion* forwardRotationIn, struct Quaternion* objectRotationInOut) {
+    if (!(flags & GrabRotationSnapToCubeNormals) || (flags & GrabRotationTurnTowardsPlayer)) {
+        return; // GrabRotationTurnTowardsPlayer takes precedence anyways if enabled
+    }
+    struct Vector3 forward, up;
+    quatMultVector(forwardRotationIn, &gForward, &forward);
+    quatMultVector(forwardRotationIn, &gUp, &up);
+    
+    int closestNormalTowards = 0, closestNormalUp = 0;
+    float closestNormalTowardsDot = 1.0f, closestNormalUpDot = -1.0f;
+    for (int i = 0; i < 6; ++i) {
+        struct Vector3 surfaceNormal;
+        quatMultVector(objectRotationInOut, &gCubeSurfaceNormals[i], &surfaceNormal);
+        
+        float dot = vector3Dot(&surfaceNormal, &forward);
+        if (dot < closestNormalTowardsDot) {
+            closestNormalTowardsDot = dot;
+            closestNormalTowards = i;
+        }
+        dot = vector3Dot(&surfaceNormal, &up);
+        if (dot > closestNormalUpDot) {
+            closestNormalUpDot = dot;
+            closestNormalUp = i;
+        }
+    }
+    struct Quaternion normalRotation;
+    quatLook(&gCubeSurfaceNormals[closestNormalTowards], &gCubeSurfaceNormals[closestNormalUp], &normalRotation);
+    quatConjugate(&normalRotation, &normalRotation);
+    quatMultiply(forwardRotationIn, &normalRotation, objectRotationInOut);
+}
+
+void grabRotationApplyTurnTowardsPlayer(const enum GrabRotationFlags flags, struct Quaternion* forwardRotationIn, struct Quaternion* objectRotationInOut) {
+    if (!(flags & GrabRotationTurnTowardsPlayer)) {
+        return;
+    }
+    *objectRotationInOut = *forwardRotationIn;
+}
+
+void grabRotationApplyUseZLookDirection(const enum GrabRotationFlags flags, struct Quaternion* lookRotationDeltaIn, struct Quaternion* grabRotationBaseInOut) {
+    if (!(flags & GrabRotationUseZLookDirection)) {
+        return;
+    }
+    struct Quaternion tmp;
+    quatMultiply(lookRotationDeltaIn, grabRotationBaseInOut, &tmp);
+    *grabRotationBaseInOut = tmp;
+}
+
+void grabRotationInitBase(const enum GrabRotationFlags flags, struct Quaternion* forwardRotationIn, struct Quaternion* objectRotationIn, struct Quaternion* grabRotationBaseOut) {
+    // modify object rotation according to flags
+    struct Quaternion objectRotation = *objectRotationIn;
+    grabRotationApplySnapToCubeNormals(flags, forwardRotationIn, &objectRotation);
+    grabRotationApplyTurnTowardsPlayer(flags, forwardRotationIn, &objectRotation);
+    
+    // untangle objectRotation from relative forwardRotation, store as grabRotationBase
+    struct Quaternion forwardRotationInverted;
+    quatConjugate(forwardRotationIn, &forwardRotationInverted);

```

Comment:
If both `GrabRotationSnapToCubeNormals` and `GrabRotationTurnTowardsPlayer` are false, then the relative `objectRotation` will be stored, such that the object isn't rotated when grabbing. I wanted to keep this as modular as possible (i.e. avoid cross dependencies between flags). But it can also be done with an extra check I guess.